Add `usa-link` component

heymatthenry opened the pull request:
> [!Note]
> This is a basic implementation of this component which is not necessarily where it will need to be for alpha status. This relates to #9 but the work in this PR does not close that issue. 

This is a basic implementation of a `usa-link` component. It supports both JS-required and a progressively-enhanced version. So a user can just include the component, like:

```html
<usa-link href="https://designsystem.digital.gov">Use USWDS!</usa-link>
```
which will render a vanilla HTML link to the shadow DOM with the right styling (additional functionality TK). Alternatively, a user can include their own link which will render as the original HTML before JS runs (or if it doesn't run at all), and will get additional functionality if/when the script executes.

```html
<usa-link>
    <a href="https://designsystem.digital.gov">Use USWDS!</a>
</usa-link>
```

In practice, that looks like this. This shows the case when JS is disabled. The first text on the page is the JS-required version, which renders as unclickable text. The second `usa-link`, with the plain HTML link as a child, still renders:
<img width="289" alt="image" src="https://github.com/uswds/uswds-next/assets/6290/7abc677b-2d82-4840-8b42-0465b56bd48e">

This is what the DOM looks like for these two versions of the component when JS doesn't run:
<img width="442" alt="image" src="https://github.com/uswds/uswds-next/assets/6290/202330a9-f1b2-40ef-be17-776f8179f808">

When JS does run, both versions are equivalent with the same styling applied to both.

heymatthenry committed:
add lit-based link; tests

heymatthenry committed:
Add basic progressively-enhanced version

heymatthenry committed:
rm extraneous files from Astro

mejiaj commented on web-components/src/components/usa-link/index.js:
```diff
@@ -0,0 +1,69 @@
+ * @csspart
+ */
+export class UsaLink extends LitElement {
+  static get properties() {
+    return {
+      /**
+       * The destination URL for the link
+       */
+      href: { type: String },
+    };
+  }
```
Curious on this approach over the examples shown in Lit Docs via [Reactive properties – Lit](https://lit.dev/docs/components/properties/).

heymatthenry replied:
The short answer is: I was just trying to get something on the screen at a time when I didn't know this API was available. But I can see the value in getting a little magic in this prop.

mejiaj replied:
Thanks for that explanation! Understandable, still getting familiar with static props.

Resolved in c3d4195.

mejiaj commented on web-components/src/components/usa-link/index.js:
```diff
@@ -0,0 +1,69 @@
+      : html`<a class="usa-link" href="${this.href}"><slot></slot></a>`;
+  }
+
+  static styles = css`
```
It would be nice to establish a pattern on where styles should live. 

Placing them above `render()` feels familiar to setting a `<link rel="stylesheet">` above HTML markup. Though I can see the benefits of having styles after for convenience while developing.

mahoneycm replied:
Is there any benefit of having css in JS like this over having a separate `styles.css` sheet and pulling it in via `unsafeCSS`?

I see a benefit of being able to write CSS in a CSS file and keeping `index.js` as JS focused as possible

heymatthenry replied:
Yeah I don't have any objections to keeping styles in separate files alongside the component code. Definitely helps maintainability. I'll factor them out here.

mejiaj commented on web-components/src/components/usa-link/index.js:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  render() {
+    return this.hasLinkChild()
+      ? html`<a class="usa-link" href="${this.href}"
+          >${this.slottedChildren}</a
+        >`
+      : html`<a class="usa-link" href="${this.href}"><slot></slot></a>`;
```
We could improve readability and make it easier to update if we separated these templates out. Like in the example here - [Composing templates – Lit](https://lit.dev/docs/components/rendering/#composing-templates)

```js
    linkChildTemplate() {
      return html`<a class="usa-link" href="${this.href}">${this.slottedChildren}</a>`
    }
    linkSlotTemplate() {
      return html`<a class="usa-link" href="${this.href}"><slot></slot></a>`
    }

    return this.hasLinkChild()
      ? this.linkChildTemplate()
      : this.linkSlotTemplate()
```

mahoneycm replied:
I'm personally a fan of this. I took a similar approach with the [card component](https://github.com/uswds/uswds-next/pull/13).

mejiaj reviewed (commented):
Looks good, I was able to test both variants successfully. As well as `tel` and `mailto` without any issues.

I was also able to override an individual link via a CSS class.

```html
  <style>
    :root {
      --theme-link-color: orange;
    }

    .james {
      --theme-link-color: green;
    }
  </style>

  <p>
    <usa-link href="http://designsystem.digital.gov">
      It's dangerous to go alone. Here, take this.
    </usa-link>
  </p>

  <p>
    <usa-link class="james">
      <a href="https://designsystem.digital.gov">It's dangerous to go alone. Here, take this.</a>
    </usa-link>
  </p>

  <p>
    <usa-link>
      <a href="https://google.com">A test link</a>
    </usa-link>
  </p>
```

Results in:
<img src="https://github.com/uswds/uswds-next/assets/3385219/8476584e-bc1c-44bd-abb2-5b20da843967" width="300" />

heymatthenry committed:
Factor styles out

heymatthenry committed:
Refactor templates

heymatthenry committed:
Add jsdoc comment

heymatthenry committed:
Use reactive prop

heymatthenry commented:
@mejiaj Thanks for your thoughtful feedback! I was totally onboard with your suggestions, and implemented them. I also added some JSDoc comments, along the lines of what's described [here](https://custom-elements-manifest.open-wc.org/analyzer/getting-started/) (not necessarily suggesting incorporating that tool unless it makes sense, but I've seen a lot of components doing docs with this format).
